src/Network/Minio/Sign/V4.hs

```diff
@@ -81,94 +84,145 @@ debugPrintSignV4Data (SignV4Data t s cr h2s ih oh sts sk) = do
-
-    authHeader = (mk "Authorization", authHeaderValue)
-
+-- AWS Sign V4 data. For normal requests (i.e. without an expiry
+-- time), the output is the list of headers to add to authenticate the
+-- request.
+--
+-- If `expiry` is not Nothing, it is assumed that a presigned request
+-- is being created. The expiry is interpreted as an integer number of
+-- seconds. The output will be the list of query-parameters to add to
+-- the request.
+signV4AtTime :: UTCTime -> ConnectInfo -> RequestInfo -> Maybe Int
+             -> SignV4Data
+signV4AtTime ts ci ri expiry =
+  let
     region = maybe (connectRegion ci) identity $ riRegion ri
-
-    scope = getScope ts region
-
-    authHeaderValue = B.concat [
-      "AWS4-HMAC-SHA256 Credential=",
-      encodeUtf8 (connectAccessKey ci), "/", scope,
-      ", SignedHeaders=", signedHeaders,
-      ", Signature=", signature
-      ]
-
-    headersToSign = getHeadersToSign headersWithDate
-
-    signedHeaders = B.intercalate ";" $ map fst headersToSign
-
-    signature = digestToBase16 $ hmacSHA256 stringToSign signingKey
-
-    signingKey = hmacSHA256RawBS "aws4_request"
-               . hmacSHA256RawBS "s3"
-               . hmacSHA256RawBS (encodeUtf8 region)
-               . hmacSHA256RawBS (awsDateFormatBS ts)
-               $ B.concat ["AWS4", encodeUtf8 $ connectSecretKey ci]
-
-    stringToSign  = B.intercalate "\n"
-      [ "AWS4-HMAC-SHA256"
-      , timeBS
+    scope = mkScope ts region
+    accessKey = toS $ connectAccessKey ci
+    secretKey = toS $ connectSecretKey ci
+
+    -- headers to be added to the request
+    datePair = ("X-Amz-Date", awsTimeFormatBS ts)
+    computedHeaders = riHeaders ri ++
+                      if isJust expiry
+                      then []
+                      else [(\(x, y) -> (mk x, y)) datePair]
+    headersToSign = getHeadersToSign computedHeaders
+    signedHeaderKeys = B.intercalate ";" $ sort $ map fst headersToSign
+
+    -- query-parameters to be added before signing for presigned URLs
+    -- (i.e. when `isJust expiry`)
+    authQP = [ ("X-Amz-Algorithm", "AWS4-HMAC-SHA256")
+             , ("X-Amz-Credential", B.concat [accessKey, "/", scope])
+             , datePair
+             , ("X-Amz-Expires", maybe "" show expiry)
+             , ("X-Amz-SignedHeaders", signedHeaderKeys)
+             ]
+    finalQP = riQueryParams ri ++
+              if isJust expiry
+              then map (\(x, y) -> (x, Just y)) authQP
+              else []
+
+    -- 1. compute canonical request
+    canonicalRequest = mkCanonicalRequest (ri {riQueryParams = finalQP})
+                       headersToSign
+
+    -- 2. compute string to sign
+    stringToSign = mkStringToSign ts scope canonicalRequest
+
+    -- 3.1 compute signing key
+    signingKey = mkSigningKey ts region secretKey
+
+    -- 3.2 compute signature
+    signature = computeSignature stringToSign signingKey
+
+    -- 4. compute auth header
+    authValue = B.concat
```

Comment:
could we use `formatBS` to make the value construction resemble AWS S3 REST API document?
something like,
`formatBS "AWS4-HMAC-SHA256 Credential={}/{}" [accessKey, scope]`

Reply:
No, formatBS requires all the args to have show instances. ByteString does not have show instance.